### **Live Project : Bangalore Tech Salary Decoder**


In [10]:
import pandas as pd
import numpy as np


In [12]:
df = pd.read_csv('/content/bangalore_tech_salaries.csv')
print("Dataset Loaded Successfully! ")

# Data inspection
print(f"Shape of the data: {df.shape} \n")
print(f"Columns of the data: {df.columns} \n")
print(f"information: {df.info()} \n")
print(f"Null values: {df.isnull().sum()} \n")
print(f"Sample data: {df.head()} \n")


Dataset Loaded Successfully! 
Shape of the data: (1015, 12) 

Columns of the data: Index(['Employee_ID', 'Role ', 'years_exp', 'Current_CTC', 'Previous_CTC',
       'Company', 'company_TYPE', 'Skills', 'Location', 'Education_Tier',
       'Joining_Year', 'Work_Mode'],
      dtype='object') 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 no

Dataset Summary:
Rows:1015
columns:12
Null Value Columns:3(Previous_CTC,Skills,Location)

In [23]:
duplicates = df.duplicated().sum()
print("Duplicate rows:", duplicates)

df.drop_duplicates(inplace=True)

print("Shape after removing duplicates:", df.shape)

Duplicate rows: 0
Shape after removing duplicates: (1000, 12)


In [24]:
df['Current_CTC']=df['Current_CTC'].str.replace('LPA','')
df['Previous_CTC']=df['Previous_CTC'].str.replace('LPA','')

df['Current_CTC']=df['Current_CTC'].str.replace('₹','')
df['Previous_CTC']=df['Previous_CTC'].str.replace('₹','')

print(df['Current_CTC'])
print("\n")
print(df['Previous_CTC'])

0            49.4
1             9.7
2       3,360,000
3           41.4 
4       3,640,000
          ...    
1010         17.6
1011        40.2 
1012         8.6 
1013         27.6
1014        20.3 
Name: Current_CTC, Length: 1000, dtype: object


0       32.4 
1         NaN
2       24.2 
3        30.3
4       30.6 
        ...  
1010     12.4
1011     26.6
1012     6.0 
1013     17.4
1014    15.4 
Name: Previous_CTC, Length: 1000, dtype: object


To convert the strings into numbers


In [28]:
# Convert Current_CTC and Previous_CTC into float (LPA)

def clean_ctc(value):

    if pd.isna(value):
        return np.nan

    value = str(value)
    value = value.replace(",", "")
    value = value.strip()

    try:
        num = float(value)

        # If salary is stored in rupees, convert it to LPA
        if num > 1000:
            num = num / 100000

        return num

    except:
        return np.nan


df["current_ctc"] = df["current_ctc"].apply(clean_ctc)
df["previous_ctc"] = df["previous_ctc"].apply(clean_ctc)

print(df[["current_ctc", "previous_ctc"]].head())
print(df[["current_ctc", "previous_ctc"]].dtypes)

   current_ctc  previous_ctc
0         49.4          32.4
1          9.7           NaN
2         33.6          24.2
3         41.4          30.3
4         36.4          30.6
current_ctc     float64
previous_ctc    float64
dtype: object


In [29]:
df.columns = (df.columns.str.strip().str.lower().str.replace(" ", "_"))

print(df.columns)

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')


In [27]:
print(df["role"].unique())

['DS' 'DevOps' 'SDE Backend' 'SDE Frontend' 'Product Lead'
 'Product Manager' 'Site Reliability Engineer' 'DA' 'UI/UX' 'SDE-Frontend'
 'UI Designer' 'ML Engineer' 'Frontend Dev' 'Data Scientist'
 'Full Stack Engineer' 'Designer' 'Product Designer' 'BA'
 'Frontend Engineer' 'BI Analyst' 'Backend Dev' 'Sr PM' 'FE' 'SDE-Backend'
 'Data Analyst' 'Business Analyst' 'Backend Developer'
 'Business Systems Analyst' 'Data Science Engineer' 'FullStack' 'BE'
 'Frontend Developer' 'Backend Engineer' 'UX Designer' 'Infra Engineer'
 'PM' 'SDE FS' 'DevOps Engineer' 'Analytics Engineer' 'Fullstack Dev'
 'SDE Full-Stack' 'SRE']


In [30]:
# Standardize role names

role_map = {

    "DS": "Data Scientist",
    "Data Science Engineer": "Data Scientist",

    "DA": "Data Analyst",
    "BI Analyst": "Data Analyst",
    "Analytics Engineer": "Data Analyst",

    "BA": "Business Analyst",
    "Business Systems Analyst": "Business Analyst",

    "PM": "Product Manager",
    "Sr PM": "Product Manager",
    "Product Lead": "Product Manager",

    "UI/UX": "UI/UX Designer",
    "Designer": "UI/UX Designer",
    "UI Designer": "UI/UX Designer",
    "UX Designer": "UI/UX Designer",
    "Product Designer": "UI/UX Designer",

    "DevOps": "DevOps Engineer",

    "SRE": "Site Reliability Engineer",
    "Infra Engineer": "Site Reliability Engineer",

    "BE": "SDE Backend",
    "Backend Dev": "SDE Backend",
    "Backend Developer": "SDE Backend",
    "Backend Engineer": "SDE Backend",
    "SDE-Backend": "SDE Backend",

    "FE": "SDE Frontend",
    "Frontend Dev": "SDE Frontend",
    "Frontend Developer": "SDE Frontend",
    "Frontend Engineer": "SDE Frontend",
    "SDE-Frontend": "SDE Frontend",

    "SDE FS": "SDE Full-Stack",
    "FullStack": "SDE Full-Stack",
    "Full Stack Engineer": "SDE Full-Stack",
    "Fullstack Dev": "SDE Full-Stack"

}

df["role_clean"] = df["role"].replace(role_map)

print(df["role_clean"].value_counts())

role_clean
SDE Backend                  114
UI/UX Designer               113
Product Manager              112
Business Analyst             108
SDE Frontend                 105
Data Analyst                 100
SDE Full-Stack                96
Data Scientist                94
Site Reliability Engineer     81
DevOps Engineer               46
ML Engineer                   31
Name: count, dtype: int64


In [31]:
print(df["company_type"].unique())

['Unicorn' 'MNC' 'Mid-size' 'early-stage' 'Early-stage' 'EARLY-STAGE'
 'mnc' 'unicorn' 'MID-SIZE' 'mid-size' 'UNICORN']


In [32]:
# Mapping company type

company_map = {
    "unicorn": "Unicorn",
    "UNICORN": "Unicorn",

    "mnc": "MNC",

    "mid-size": "Mid-size",
    "MID-SIZE": "Mid-size",

    "early-stage": "Early-stage",
    "EARLY-STAGE": "Early-stage"
}

df["company_type_clean"] = (
    df["company_type"]
    .replace(company_map)
)

print(df["company_type_clean"].value_counts())

company_type_clean
MNC            342
Mid-size       257
Unicorn        207
Early-stage    194
Name: count, dtype: int64


In [33]:
print(df["education_tier"].unique())

['Tier-1' '2' 'Tier-2' 'Tier 1' 'Tier 2' 'T3' 'Tier-3' 'T2' '1' 'T1'
 'Tier 3' '3']


In [34]:
# Standardize education tier

education_map = {
    "Tier-1": "Tier 1",
    "Tier-2": "Tier 2",
    "Tier-3": "Tier 3",
    "T1": "Tier 1",
    "T2": "Tier 2",
    "T3": "Tier 3",
    "1": "Tier 1",
    "2": "Tier 2",
    "3": "Tier 3"
}

df["education_tier_clean"] = (
    df["education_tier"]
    .astype(str)
    .replace(education_map)
)

print(df["education_tier_clean"].value_counts())

education_tier_clean
Tier 2    500
Tier 3    306
Tier 1    194
Name: count, dtype: int64


In [35]:
# Fill missing values in Skills and Location

df["skills"] = df["skills"].fillna("Not Available")
df["location"] = df["location"].fillna("Unknown")

# Keeping Previous_CTC as NaN because freshers don't have previous salary

print(df.isnull().sum())

employee_id               0
role                      0
years_exp                 0
current_ctc               0
previous_ctc            194
company                   0
company_type              0
skills                    0
location                  0
education_tier            0
joining_year              0
work_mode                 0
role_clean                0
company_type_clean        0
education_tier_clean      0
dtype: int64


In [36]:
print("\nData Types\n")
print(df.dtypes)

print("\nRole Count\n")
print(df["role_clean"].value_counts())

print("\nCompany Type Count\n")
print(df["company_type_clean"].value_counts())

print("\nEducation Tier Count\n")
print(df["education_tier_clean"].value_counts())


Data Types

employee_id              object
role                     object
years_exp                 int64
current_ctc             float64
previous_ctc            float64
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
company_type_clean       object
education_tier_clean     object
dtype: object

Role Count

role_clean
SDE Backend                  114
UI/UX Designer               113
Product Manager              112
Business Analyst             108
SDE Frontend                 105
Data Analyst                 100
SDE Full-Stack                96
Data Scientist                94
Site Reliability Engineer     81
DevOps Engineer               46
ML Engineer                   31
Name: count, dtype: int64

Company Type Count

company_type_clean
MNC            342
Mid-size       257
Un

In [39]:
# Q3.1 - Calculate median, mean, minimum and maximum salary for each role

role_salary = (
    df.groupby("role_clean")["current_ctc"]
    .agg(["median", "mean", "min", "max"])
    .sort_values(by="median", ascending=False)
)
print(role_salary)
print("\n")
highest = role_salary.index[0]
lowest = role_salary.index[-1]

print("\nHighest Paying Role :", highest)
print("Lowest Paying Role :", lowest)

                           median       mean   min   max
role_clean                                              
Product Manager             31.30  34.131250  10.8  80.1
ML Engineer                 27.50  30.335484  10.0  64.5
Data Scientist              24.15  26.670213  10.9  75.6
Site Reliability Engineer   23.30  23.481481   8.8  55.0
SDE Full-Stack              22.35  25.409375   8.9  71.7
SDE Frontend                21.20  22.193333   6.7  84.4
SDE Backend                 21.10  23.284210   7.9  55.1
Business Analyst            19.90  21.873148   6.8  52.7
DevOps Engineer             19.45  22.891304   9.7  60.3
UI/UX Designer              18.90  20.968142   6.2  63.3
Data Analyst                16.90  17.874000   5.2  43.4



Highest Paying Role : Product Manager
Lowest Paying Role : Data Analyst


### Insight (Q3.1)

The role with the highest median salary is Product Manager, while Data Analyst has the lowest median salary.
The difference between mean and median shows whether some roles have high salary outliers.

In [42]:
# Q3.2 - Compare salary based on experience for SDE Backend

backend = df[df["role_clean"] == "SDE Backend"].copy()

# Create experience bands
bins = [-1, 1, 3, 5, 100]
labels = ["0-1 Years", "2-3 Years", "4-5 Years", "6+ Years"]

backend["exp_band"] = pd.cut(
    backend["years_exp"],
    bins=bins,
    labels=labels
)

# Median salary of each experience band
exp_salary = backend.groupby("exp_band", observed=False)["current_ctc"].median()

print("Median Salary by Experience Band\n")
print(exp_salary)

print("\nGrowth Percentage\n")

previous = None

for band in exp_salary.index:

    salary = exp_salary[band]

    if previous is None:
        print(f"{band} : {salary:.2f} LPA")
    else:
        growth = ((salary - previous) / previous) * 100
        print(f"{band} : {salary:.2f} LPA (+{growth:.2f}%)")

    previous = salary

Median Salary by Experience Band

exp_band
0-1 Years    11.65
2-3 Years    20.00
4-5 Years    25.85
6+ Years     40.40
Name: current_ctc, dtype: float64

Growth Percentage

0-1 Years : 11.65 LPA
2-3 Years : 20.00 LPA (+71.67%)
4-5 Years : 25.85 LPA (+29.25%)
6+ Years : 40.40 LPA (+56.29%)


In [44]:
# Q3.4 - Compare SDE Backend salary across company types

backend = df[df["role_clean"] == "SDE Backend"]

company_salary = (
    backend.groupby("company_type_clean")["current_ctc"]
    .median()
)

print(company_salary)

unicorn = company_salary["Unicorn"]
mnc = company_salary["MNC"]

premium = ((unicorn - mnc) / mnc) * 100

print("\nUnicorn Premium over MNC =", round(premium,2),"%")

company_type_clean
Early-stage    18.60
MNC            20.45
Mid-size       19.50
Unicorn        27.35
Name: current_ctc, dtype: float64

Unicorn Premium over MNC = 33.74 %


In [45]:
# Q3.5 - Find employees earning below their group's median salary

bins = [0,1,3,5,50]
labels = ["0-1 Years","2-3 Years","4-5 Years","6+ Years"]

df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=bins,
    labels=labels,
    include_lowest=True
)

group_size = (
    df.groupby(
        ["role_clean","company_type_clean","exp_band"]
    )["current_ctc"]
    .transform("count")
)

group_median = (
    df.groupby(
        ["role_clean","company_type_clean","exp_band"]
    )["current_ctc"]
    .transform("median")
)

underpaid = df[group_size >= 10].copy()

underpaid["group_median"] = group_median[group_size >= 10]

underpaid["gap"] = (
    underpaid["current_ctc"] -
    underpaid["group_median"]
)

underpaid = underpaid.sort_values("gap").head(10)

print(
    underpaid[
        [
            "employee_id",
            "role_clean",
            "company_type_clean",
            "years_exp",
            "current_ctc",
            "group_median",
            "gap"
        ]
    ]
)

    employee_id        role_clean company_type_clean  years_exp  current_ctc  \
904     BLR0925       SDE Backend                MNC          4         19.4   
250     BLR0072       SDE Backend                MNC          4         19.5   
249     BLR0733  Business Analyst                MNC          7         31.1   
728     BLR0301   Product Manager                MNC          2         24.7   
386     BLR0453  Business Analyst                MNC          6         32.1   
753     BLR0611   Product Manager                MNC          2         25.1   
711     BLR0451      SDE Frontend                MNC          4         19.5   
688     BLR0203       SDE Backend                MNC          4         21.3   
858     BLR0033      SDE Frontend                MNC          4         19.6   
538     BLR0486   Product Manager                MNC          2         25.7   

     group_median   gap  
904         27.30 -7.90  
250         27.30 -7.80  
249         38.50 -7.40  
728         31.

/tmp/ipykernel_3638/1709492594.py:14: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(
/tmp/ipykernel_3638/1709492594.py:21: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(


In [46]:
# ==============================
# TASK 4 - FINAL PRINTED REPORT
# ==============================

print("=" * 60)
print("           BANGALORE TECH SALARY DECODER")
print("      Built by Vaishanavi Kashid | The Unlox Academy")
print("=" * 60)

print("Dataset : 1000 Bengaluru Tech Professionals")
print("Period  : 2024 Employment Snapshot")

# ------------------------------------
# Median CTC by Role
# ------------------------------------

print("\n---- MEDIAN CTC BY ROLE (LPA) ----")

for role, salary in role_salary["median"].items():

    bar = "█" * int(salary // 2)

    print(f"{role:<28} {bar:<18} {salary:.2f}")

# ------------------------------------
# SDE Backend Experience
# ------------------------------------

print("\n---- SDE BACKEND CTC BY EXPERIENCE ----")

previous = None

for band in exp_salary.index:

    salary = exp_salary[band]

    if previous is None:

        print(f"{band:<12}: {salary:.2f} LPA")

    else:

        growth = ((salary - previous) / previous) * 100

        print(f"{band:<12}: {salary:.2f} LPA (+{growth:.2f}%)")

    previous = salary

# ------------------------------------
# Skill Premium
# ------------------------------------

print("\n---- SKILL PREMIUM FOR SDEs ----")

skills = ["AWS", "ML", "System Design", "Kubernetes"]

print(f"{'Skill':<18}{'With':<12}{'Without':<12}{'Premium'}")

sde = df[df["role_clean"].isin([
    "SDE Backend",
    "SDE Frontend",
    "SDE Full-Stack"
])]

for skill in skills:

    with_skill = sde[
        sde["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc"].median()

    without_skill = sde[
        ~sde["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc"].median()

    premium = ((with_skill - without_skill) / without_skill) * 100

    print(f"{skill:<18}{with_skill:<12.2f}{without_skill:<12.2f}+{premium:.2f}%")

# ------------------------------------
# Company Premium
# ------------------------------------

print("\n---- COMPANY TYPE PREMIUM (SDE Backend) ----")

highest = company_salary.max()

for company, salary in company_salary.items():

    diff = ((highest - salary) / highest) * 100

    if company == company_salary.idxmax():

        print(f"{company:<15}{salary:.2f} LPA  <-- Highest")

    else:

        print(f"{company:<15}{salary:.2f} LPA (-{diff:.2f}%)")

# ------------------------------------
# Top 5 Underpaid Employees
# ------------------------------------

print("\n---- TOP 5 MOST UNDERPAID PROFESSIONALS ----")

top5 = underpaid.head(5)

for _, row in top5.iterrows():

    print(
        f"{row['employee_id']}  "
        f"{row['role_clean']:<18}"
        f"{row['company_type_clean']:<12}"
        f"{row['years_exp']} yrs   "
        f"Gap : {row['gap']:.2f} LPA"
    )

print("=" * 60)
print("                END OF REPORT")
print("=" * 60)

           BANGALORE TECH SALARY DECODER
      Built by Vaishanavi Kashid | The Unlox Academy
Dataset : 1000 Bengaluru Tech Professionals
Period  : 2024 Employment Snapshot

---- MEDIAN CTC BY ROLE (LPA) ----
Product Manager              ███████████████    31.30
ML Engineer                  █████████████      27.50
Data Scientist               ████████████       24.15
Site Reliability Engineer    ███████████        23.30
SDE Full-Stack               ███████████        22.35
SDE Frontend                 ██████████         21.20
SDE Backend                  ██████████         21.10
Business Analyst             █████████          19.90
DevOps Engineer              █████████          19.45
UI/UX Designer               █████████          18.90
Data Analyst                 ████████           16.90

---- SDE BACKEND CTC BY EXPERIENCE ----
0-1 Years   : 11.65 LPA
2-3 Years   : 20.00 LPA (+71.67%)
4-5 Years   : 25.85 LPA (+29.25%)
6+ Years    : 40.40 LPA (+56.29%)

---- SKILL PREMIUM FOR SDEs -